packages/@webex/plugin-meetings/src/multistream/remoteMediaManager.ts

```diff
@@ -0,0 +1,891 @@
+    {id: 'stage-4', size: 'medium', csi: undefined},
+  ],
+};
+
+/**
+ * Default configuration:
+ * - uses 3 audio streams
+ * - prefers active speakers with live video (e.g. are not audio only or video muted) over active speakers without live video
+ * - has a few layouts defined, including 1 that contains remote screen share (ScreenShareView)
+ */
+export const DefaultConfiguration: Configuration = {
+  audio: {
+    numOfActiveSpeakerStreams: 3,
+  },
+  video: {
+    preferLiveVideo: true,
+    initialLayoutId: 'AllEqual',
+
+    layouts: {
+      AllEqual: AllEqualLayout,
+      OnePlusFive: OnePlusFiveLayout,
+      Single: SingleLayout,
+      Stage: Stage2x2With6ThumbnailsLayout,
+      ScreenShareView: RemoteScreenShareWithSmallThumbnailsLayout,
+    },
+  },
+  screenShare: {
+    audio: true,
+    video: true,
+  },
+};
+
+export enum Event {
+  // events for audio streams
+  AudioCreated = 'AudioCreated',
+  ScreenShareAudioCreated = 'ScreenShareCreated',
+
+  // events for video streams
+  VideoLayoutChanged = 'VideoLayoutChanged',
+}
+
+export interface VideoLayoutChangedEventData {
+  layoutId: LayoutId;
+  activeSpeakerVideoPanes: {
+    [key: PaneGroupId]: RemoteMediaGroup;
+  };
+  memberVideoPanes: {[key: PaneId]: RemoteMedia};
+  screenShareVideo?: RemoteMedia;
+}
+export interface Events extends EventMap {
+  // audio
+  [Event.AudioCreated]: (audio: RemoteMediaGroup) => void;
+  [Event.ScreenShareAudioCreated]: (screenShareAudio: RemoteMedia) => void;
+
+  // video
+  [Event.VideoLayoutChanged]: (data: VideoLayoutChangedEventData) => void;
+}
+
+/**
+ * A helper class that manages all remote audio/video streams in order to achieve a predefined set of layouts.
+ * It also creates a fixed number of audio streams and these don't change during the meeting.
+ *
+ * Things that RemoteMediaManager does:
+ * - owns the receive slots (creates them when needed, and re-uses them when switching layouts)
+ * - constructs appropriate RemoteMedia and RemoteMediaGroup objects and sends appropriate mediaRequests
+ */
+export class RemoteMediaManager extends EventsScope {
+  private config: Configuration;
+
+  private receiveSlotManager: ReceiveSlotManager;
+
+  private mediaRequestManagers: {
+    audio: MediaRequestManager;
+    video: MediaRequestManager;
+  };
+
+  private currentLayout?: VideoLayout;
+
+  private slots: {
+    audio: ReceiveSlot[];
+    screenShareAudio?: ReceiveSlot;
+    screenShareVideo?: ReceiveSlot;
+    video: {
+      unused: ReceiveSlot[];
+      activeSpeaker: ReceiveSlot[];
+      receiverSelected: ReceiveSlot[];
+    };
+  };
+
+  private media: {
+    audio?: RemoteMediaGroup;
+    video: {
+      activeSpeakerGroups: {
+        [key: PaneGroupId]: RemoteMediaGroup;
+      };
+      memberPanes: {[key: PaneId]: RemoteMedia};
+    };
+  };
+
+  private receiveSlotAllocations: {
+    activeSpeaker: {[key: PaneGroupId]: {slots: ReceiveSlot[]}};
+    receiverSelected: {[key: PaneId]: ReceiveSlot};
```

Comment:
Why have different naming system for slots & media activeSpeaker/receiverSelected vs activeSpeakerGroups/memberPanes it seems unnecessarily confusing

Reply:
Hmm not sure if making them all have same names will be better. The slots are more low level and we just operate on active speaker slots or receiver selected slots hence the names, while for media, this is higher level and the active speaker media object actually contains instances of `RemoteMediaGroup` hence the "Groups" in `this.media.video.activeSpeakerGroups` name. Also I was trying to align the media names with the public API of the VideoLayout interface, which has "activeSpeakerVideoPaneGroups and memberVideoPanes" - the media has the same names just with "Video" removed, because it's already under `this.media.video` so I felt that "Video" was redundant.

Reply:
ok